python/python/bystro/cli/auth.py

```diff
@@ -0,0 +1,242 @@
+import argparse
+import os
+
+import requests
+
+from msgspec import Struct, json as mjson
+
+DEFAULT_DIR = os.path.expanduser("~/.bystro")
+STATE_FILE = "bystro_authentication_token.json"
+
+
+class SignupResponse(Struct, rename="camel"):
+    """
+    The response body for signing up for Bystro.
+
+    Attributes
+    ----------
+    access_token : str
+        The access token, which authorizes further API requests
+    """
+
+    access_token: str
+
+
+class LoginResponse(Struct, rename="camel"):
+    """
+    The response body for logging in to Bystro.
+
+    Attributes
+    ----------
+    access_token : str
+        The access token, which authorizes further API requests
+    """
+
+    access_token: str
+
+
+class CachedAuth(Struct, rename="camel"):
+    """
+    The authentication state.
+
+    Attributes
+    ----------
+    email : str
+        The email of the user.
+    access_token : str
+        The access token, which authorizes further API requests
+    url : str
+        The url of the Bystro server.
+    """
+
+    email: str
+    access_token: str
+    url: str
+
+
+def _fq_host(args: argparse.Namespace) -> str:
+    """
+    Returns the fully qualified host, e.g. https://bystro-dev.emory.edu:443
+
+    Parameters
+    ----------
+    args : argparse.Namespace
+        The arguments passed to the command.
+
+    Returns
+    -------
+    str
+        The fully qualified host.
+    """
+    return f"{args.host}:{args.port}"
+
+
+def load_state(bystro_credentials_dir: str = DEFAULT_DIR) -> CachedAuth | None:
+    """
+    Loads the authentication state from the state directory.
+
+    Parameters
+    ----------
+    bystro_credentials_dir : str
+        The directory where the authentication state is saved.
+
+    Returns
+    -------
+    CachedAuth | None
+        The authentication state, or None if the state file doesn't exist.
+    """
+    path = os.path.join(bystro_credentials_dir, STATE_FILE)
+
+    if os.path.exists(path):
+        with open(path, "r", encoding="utf-8") as f:
+            return mjson.decode(f.read(), type=CachedAuth)
+
+    return None
+
+
+def save_state(data: CachedAuth, bystro_credentials_dir: str = DEFAULT_DIR, print_result=True) -> None:
```

Comment:
This is again duplicated. All duplicated code should be removed. The only functions here should be those that takes argsparse.Namespace, and call the api/auth.py function